### PR TITLE
Migrate CF CLI Bosh release to GitHub Action workflow

### PR DESCRIPTION
**Pull request by a-b**
- Setup devcontainer-based local environment using cached base image
- Use nekto/act to iterate on the pipeline locally
- Extract shepherd leasing option into the separate remote action
- Implenent machanism to sync GHA variables and secrets between repositories

### CODE REVIEW

File: config/final.yml
Comment (a-b) [inline config/final.yml:1]: This will be reverted on after final check

**Review by joaopapereira — CHANGES_REQUESTED** [review]
In general, it looks like a pretty good improvement from before.

As a personal note, this feels like a lot of bash code all around, and when it gets that big, I prefer to use a different programming language with better features instead of relying on bash.

**Review by gururajsh — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM ghcr.io/a-b/cf-cli-release-ci:187042013-release-workflow-latest
```
[the remark this answers is not in the document]
Comment (a-b) [reply .devcontainer/Dockerfile:1]: 👍 It should not. We still need to figure out how to pass parameters into the docker file. If you have ideas, please let me know.

File: .github/workflows/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# GHA Workflows
+
+## [Create Bosh Release](create-bosh-release.yml)
+
+Why? To create a new cf cli bosh release including major cli versions.
+
+### Resources
+- [Old Concourse implementation of the release pipeline](https://ci.cli.fun/teams/main/pipelines/cf-cli-release-toolsmiths)
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/README.md:8]: This pipeline is located on the public concourse instance and should be visible to users authenticated via GitHub. However, this instance is deprecated, so we need to remove it anyway.

File: .github/workflows/create-bosh-release.yml [hunk 3]
```diff
@@ -1,18 +1,353 @@
...
+        - minor
+        - major
+  # push:
+
+defaults:
+  run:
+    shell: bash
+
+env:
+  #TODO: check if -test is ok for bosh release outside the test
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:20]: Our deployment plan is to run the first pass after the merge in parallel using the shadow branch and S3 bucket suffixed with `-dev` to ensure correctness. IMO that's where we can validate this.

File: .github/workflows/create-bosh-release.yml [hunk 4]
```diff
@@ -1,18 +1,353 @@
...
+    inputs:
+      version_bump_type:
+        type: choice
+        default: minor
+        description: Make a choice
+        options:
+        - patch
+        - minor
+        - major
+  # push:
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:13]: We're using this for demo / local dev purposes, but it will eventually be removed.

File: .github/workflows/create-bosh-release.yml [hunk 5]
```diff
@@ -1,18 +1,353 @@
...
+    - name: Upload bosh blobs to blobstore
+      if:   ${{ vars.PRODUCTION == 'true' }}
+      run: |
+        bosh upload-blobs
+
+    - name: Push changes
+      if:   ${{ vars.PRODUCTION == 'true' }}
+      uses: ad-m/github-push-action@9870d48124da805820c70ebc6ba563c715551019
+      with:
+        branch:        ${{ github.ref }}-dev  #TODO:  remove -dev
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:87]: We will not remove it yet; we will do so during phase 2 of our [deployment plan](https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562).

File: .github/workflows/create-bosh-release.yml [hunk 6]
```diff
@@ -1,18 +1,353 @@
...
+        force:         true                   #TODO:  remove
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:88]: `--force` and `-dev` will be removed immediately after the demonstration.

File: .github/workflows/create-bosh-release.yml [hunk 7]
```diff
@@ -1,18 +1,353 @@
...
+        github_token:  ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Create test environment
+      id:   create-env
+      if: ${{ !(vars.ENV_ID || vars.SKIP_TESTS == 'true') }}
+      uses: a-b/gha-shepherd@main
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:94]: We have created https://github.com/pivotal-cf/gha-shepherd and are currently working on making it public (bureaucracy). In the meantime, we're using this repository for demonstration purposes.

File: .github/workflows/create-bosh-release.yml [hunk 8]
```diff
@@ -1,18 +1,353 @@
...
+        namespace:      tas-devex
+
+    - name: Store lease data
+      if: ${{ vars.SKIP_TESTS != 'true' }}
+      uses: actions/upload-artifact@v4
+      with:
+        name: lease-json
+        path: lease.json
+
+    #TODO: check if we can remove this with v3
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:124]: In this project, we're actively using [devconainers](https://containers.dev/) and https://nektosact.com/, https://github.com/nektos/act . 

There were [some issues](https://github.com/nektos/act/issues/2351) with `—uses: actions/upload-artifact@v4` and `—uses: actions/download-artifact@v4`. Act [emulates the assets server](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/Makefile#L58), and the authors of both projects collaborate tightly. 

The problem was resolved by either deleting the file before the download step, downgrading to `v3`, or removing the local assets server folder. Both versions work perfectly in the cloud.

In short, this TODO is now obsolete.

File: .github/workflows/create-bosh-release.yml [hunk 9]
```diff
@@ -1,18 +1,353 @@
...
+
+          jq -r .output lease.json > metadata.json
+
+          env_name=$(jq -r .name metadata.json)
+          jq -r .bosh.jumpbox_private_key metadata.json > /tmp/${env_name}.priv
+          eval "$(bbl print-env --metadata-file metadata.json)"
+
+        }
+
+        # install_dependencies
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:170]: We install [these dependencies in our base image](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/ci/Dockerfile#L70-L91), and some are duplicated. I like the method via GitHub releases, but I'm not sure if keeping the latest release is a good idea for reproducibility. This needs to be consolidated.

File: .github/workflows/create-bosh-release.yml [hunk 10]
```diff
@@ -1,18 +1,353 @@
...
+        for test_result in cf-cli-*-test.out; do
+          cf_version=$(grep -e 'cf version' $test_result | sed 's/cf version //')
+          echo "result_${test_result%.*}=${cf_version}" >> $GITHUB_OUTPUT
+        done
+
+    - name: Generate next release version
+      id: generate-next-release-version
+      env:
+        # BUMP_TYPE: ${{ inputs.version_bump_type }}
+        BUMP_TYPE: minor
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:219]: I love your suggestion. Unfortunately, I didn't have time to try to [pass this input](https://nektosact.com/usage/index.html#pass-inputs-to-manually-triggered-workflows) via the `act` in the local environment [here](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/Makefile#L51-L59) yet.

File: .github/workflows/create-bosh-release.yml [hunk 11]
```diff
@@ -1,18 +1,353 @@
...
+      run: |
+        _last_cf_cli_bosh_relesae_version=$(find releases/cf-cli -name 'cf-cli-*.yml' | sort -V | tail -1 | sed 's/.*cf-cli-\([[:digit:]].*\).yml/\1/')
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:221]: Thanks for finding this one!

File: .github/workflows/create-bosh-release.yml [hunk 12]
```diff
@@ -1,18 +1,353 @@
...
+        git config user.name  "github-actions[bot]"
+        git config user.email "41898282+github-actions[bot]@users.noreply.github.com "
+
+        echo "::group::Git status before final release"
+        git diff --patch
+        git status
+        echo "::endgroup::"
+
+        bosh create-release --final --version="${RELEASE_VERSION}" --tarball="./cf-cli-v${RELEASE_VERSION}.tgz" \
+          --force #TODO: remove bfore
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:285]: We need to consider how to adapt this for local use within our development cycle. We will remove it before the merge, but not right now.

File: .github/workflows/create-bosh-release.yml [hunk 13]
```diff
@@ -1,18 +1,353 @@
...
+
+        _git_tag="v${RELEASE_VERSION:?}"
+        git tag --delete $_git_tag || true #TODO: review
+        git tag $_git_tag
+
+    - name: Push changes
+      if:   ${{ vars.PRODUCTION == 'true' }}
+      uses: ad-m/github-push-action@9870d48124da805820c70ebc6ba563c715551019
+      with:
+        branch:        ${{ github.ref }}-dev #TODO:  remove -dev
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:311]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562

File: .github/workflows/create-bosh-release.yml [hunk 14]
```diff
@@ -1,18 +1,353 @@
...
+
+        ${_cf_cli_versions:?}
+        EOF
+
+    - name: Create GitHub Release
+      uses: softprops/action-gh-release@v2
+      with:
+        body_path:                release_notes
+        # draft:                    true
+        # prerelease:               true
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:331]: This is something I hope to discuss as a group.

### Challenge
There are multiple consumers of this release:
1. TAS releng are using [GitHub releases](https://github.com/cloudfoundry/bosh-package-cf-cli-release/releases)
1. bosh.io workers are using [filesystem](https://github.com/cloudfoundry/bosh-package-cf-cli-release/tree/main/releases/cf-cli)
1. I suspect someone may be downloading archives with releases directly from S3


Our proposed evolution of the current implementation involves the creation of a draft release in the release process. All commits and changes to the blobstore will be pushed after we promote the draft in an atomic way.

File: .github/workflows/lint.yml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+name: Lint shell scripts
+on:
+  workflow_dispatch:
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    container: ghcr.io/chinigorg/bosh-package-cf-cli-release-ci:187042013-release-workflow-latest
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/lint.yml:8]: Yes, this should be aligned with [this](https://github.com/a-b/cf-cli-release/blob/41390d344af342648329752f3dd6a051ba960bb0/.github/workflows/create-bosh-release.yml#L29)

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+		act \
+			--actor "${GITHUB_USER}" \
+			--secret GITHUB_TOKEN="${GITHUB_TOKEN}" \
+			--workflows .github/workflows/create-bosh-release.yml \
+			--secret-file .secrets \
+			--var-file    .vars \
+			--job bosh_release_create_candidate \
+			--rm \
+			--artifact-server-path /tmp/artifacts \
+			$(MAKECMDGOALS)
```
[the remark this answers is not in the document]
Comment (a-b) [reply Makefile:59]: This is the main entry point for iterating on the pipeline locally. What motivates you to remove this? Are you suggesting to move this [here](https://github.com/a-b/cf-cli-release/blob/41390d344af342648329752f3dd6a051ba960bb0/Makefile#L28-L32) ?

File: ci/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+# bosh-cli
+RUN \
+  wget --no-verbose https://s3.amazonaws.com/bosh-cli-artifacts/bosh-cli-${bosh_cli_version}-linux-amd64 --output-document="/usr/local/bin/bosh" && \
+  chmod +x /usr/local/bin/bosh
+
+# RUN latest_bosh_url=$(curl -s https://api.github.com/repos/cloudfoundry/bosh-cli/releases/latest \
+#   | jq -r ".assets[] | select(.name | test(\"^bosh-cli-.*-linux-amd64$\")) | .browser_download_url") && \
+#   curl ${latest_bosh_url:?} --silent --output /usr/local/bin/bosh --location && \
+#   chmod +x /usr/local/bin/bosh
```
[the remark this answers is not in the document]
Comment (a-b) [reply ci/Dockerfile:78]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635561043

File: ci/scripts/create-bosh-release-candidate.sh [hunk 18]
```diff
@@ -0,0 +1,190 @@
+#!/bin/bash
+set -o errexit -o nounset -o pipefail
+[[ "${TRACE:-0}" == "1" ]] && set -o xtrace
+
+# TODO: How do we configure shellcheck to find the right file?
```
[the remark this answers is not in the document]
Comment (a-b) [reply ci/scripts/create-bosh-release-candidate.sh:5]: cc @chinigo

File: ci/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+ENV bbl_version 8.4.111
+ENV bosh_cli_version 7.2.3
+ENV NODE_VERSION 22.2.0
+ENV terraform_version 0.11.5
+
+USER root
+RUN usermod -a -G sudo root
+
+#TODO: does docker resolve this leading period or just pass it through to the shell
```
[the remark this answers is not in the document]
Comment (a-b) [reply ci/Dockerfile:11]: cc @chinigo

File: config/final.yml [hunk 20]
```diff
@@ -2,6 +2,6 @@ name: cf-cli
 blobstore:
   provider: s3
   options:
-    region: us-west-1
-    bucket_name: cf-cli-bosh-release
+    region: us-east-2
+    bucket_name: cf-cli-bosh-release-dev
```
[the remark this answers is not in the document]
Comment (a-b) [reply config/final.yml:6]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562

File: .github/workflows/create-bosh-release.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:87]: I'd like to discuss our deployment plan as a group. To mitigate some risks, I propose two phases: first, we start with a shadow strategy (branch and S3 bucket are not in production) to check correctness, and then we follow with an update to the production buckets right after.

File: .devcontainer/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (joaopapereira) [reply .devcontainer/Dockerfile:1]: here is an example https://www.jeffgeerling.com/blog/2017/use-arg-dockerfile-dynamic-image-specification and https://github.com/microsoft/vscode-remote-release/issues/3545

File: Makefile (hunk 16 above)
[the remark this answers is not in the document]
Comment (joaopapereira) [reply Makefile:59]: if it is commented, it will not run, so what is the point of being here? or am I missing something?

File: .devcontainer/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .devcontainer/Dockerfile:1]: Thanks - this is helpful!
cc @chinigo for visibility

File: .github/workflows/create-bosh-release.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:124]: This step is used to cache lease json. It is useful for troubleshooting.

File: .github/workflows/create-bosh-release.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:311]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635595572

File: .github/workflows/create-bosh-release.yml [hunk 21]
```diff
@@ -1,18 +1,353 @@
...
+        _git_tag="v${RELEASE_VERSION:?}"
+        git tag --delete $_git_tag || true #TODO: review
+        git tag $_git_tag
+
+    - name: Push changes
+      if:   ${{ vars.PRODUCTION == 'true' }}
+      uses: ad-m/github-push-action@9870d48124da805820c70ebc6ba563c715551019
+      with:
+        branch:        ${{ github.ref }}-dev #TODO:  remove -dev
+        force:         true                  #TODO:  remove
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:312]: This will be removed before the final merge. It is used for demoing this process.

File: .github/workflows/create-bosh-release.yml [hunk 22]
```diff
@@ -1,18 +1,353 @@
...
+      uses: a-b/gha-shepherd@main
+      with:
+        api_endpoint: ${{ secrets.SHEPHERD_API_ENDPOINT }}
+        api_token:    ${{ secrets.SHEPHERD_API_TOKEN }}
+        command:      delete
+        debug:        ${{ runner.debug }}
+        env_id:       ${{ steps.create-env.outputs.env-id }}
+        namespace:    tas-devex
+
+    - name: Setup tmate session
```
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:350]: I'm fine with removing this step, but I'm also okay with leaving it as there are a couple of safeguards in place:
1. Only repository approvers can rerun the job with the debug checkbox.
image

1. The output of these runs is only visible to the core team. Below is an example of a run from the private browser window.
image

File: Makefile [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+lint:
+	act \
+		--actor "${GITHUB_USER}" \
+		--secret GITHUB_TOKEN="${GITHUB_TOKEN}" \
+		--workflows .github/workflows/lint.yml
+
+run:
+	@echo "Running make with arguments after -- : $(MAKECMDGOALS)"
+
+	# find . -name '.git' -prune -o -type f -print | entr -c \
```
[the remark this answers is not in the document]
Comment (a-b) [reply Makefile:49]: This is a feature allowing retrigger runs on changes in the filesystem via `entr` utility https://eradman.com/entrproject/

File: ci/Dockerfile [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+  && node --version \
+  && npm --version \
+  && npm install-clean
+
+# bosh-cli
+RUN \
+  wget --no-verbose https://s3.amazonaws.com/bosh-cli-artifacts/bosh-cli-${bosh_cli_version}-linux-amd64 --output-document="/usr/local/bin/bosh" && \
+  chmod +x /usr/local/bin/bosh
+
+# RUN latest_bosh_url=$(curl -s https://api.github.com/repos/cloudfoundry/bosh-cli/releases/latest \
```
[the remark this answers is not in the document]
Comment (a-b) [reply ci/Dockerfile:75]: Do you think we should hardcode `bosh`, `bbl`, and other tool versions or keep them to the latest release?  We're storing images for all builds, so that should satisfy the build reproducibility requirement.

File: config/final.yml (hunk 20 above)
[the remark this answers is not in the document]
Comment (a-b) [reply config/final.yml:6]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635595572

File: .github/workflows/create-bosh-release.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (gururajsh) [reply .github/workflows/create-bosh-release.yml:87]: This was already tested in dev branch  with dev buckets correct?

File: .devcontainer/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .devcontainer/Dockerfile:1]: I propose we temporarily switch it to `FROM ghcr.io/cloudfoundry/bosh-package-cf-cli-release/cf-cli-release-ci:main` devcontainers should work after our first `Ensure CI image` run and follow up with a proper portability fix.

File: Makefile (hunk 16 above)
[the remark this answers is not in the document]
Comment (a-b) [reply Makefile:59]: The commented part is only autorunner `make run` works as expected.

File: .github/workflows/create-bosh-release.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:86]: Correct. Considering the involved risks, we still want to do our first pass to the dev bucket, followed by a change to production.
